### PR TITLE
ENH: Implement StringArray.min / max

### PR DESCRIPTION
**Pull request by dsaxton**
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Using the new masked reductions from @jorisvandenbossche to implement these for StringArray. Part of https://github.com/pandas-dev/pandas/issues/31746 but doesn't close because we're not adding sum here.

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for looking at this!

File: pandas/core/arrays/string_.py [hunk 1]
```diff
@@ -282,8 +283,23 @@ def astype(self, dtype, copy=True):
         return super().astype(dtype, copy)
 
     def _reduce(self, name, skipna=True, **kwargs):
+        if name in ["min", "max"]:
+            return getattr(self, name)(skipna=skipna, **kwargs)
+
         raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
 
+    def min(self, axis=None, out=None, keepdims=False, skipna=True):
```
[the remark this answers is not in the document]
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: Updated, assuming by test with numpy you mean the validation functions in /compat/numpy/function?

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: Yes, indeed, those validation functions can be used to check additional keywords. 

In addition, we should also test this in the tests (so test that np.min(a) works, because that is the only reason we would add those keywords)

**Comment by dsaxton** [comment]
@jorisvandenbossche What do you think of implementing these reductions at the PandasArray or BaseMaskedArray level (I figure they should probably be used at least for IntegerArray anyways, but if so then might as well make them inheritable by other masked arrays, e.g., BooleanArray)?

File: pandas/core/arrays/string_.py [hunk 2]
```diff
@@ -282,8 +284,25 @@ def astype(self, dtype, copy=True):
         return super().astype(dtype, copy)
 
     def _reduce(self, name, skipna=True, **kwargs):
+        if name in ["min", "max"]:
+            return getattr(self, name)(skipna=skipna, **kwargs)
+
         raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
 
+    def min(self, skipna=True, **kwargs):
+        nv.validate_min((), kwargs)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py:293]: Can you add a `if kwargs` so we only call this if there are actually kwargs (which avoids the overhead of this there are actually no kwargs

File: pandas/core/arrays/string_.py [hunk 3]
```diff
@@ -282,8 +284,25 @@ def astype(self, dtype, copy=True):
         return super().astype(dtype, copy)
 
     def _reduce(self, name, skipna=True, **kwargs):
+        if name in ["min", "max"]:
+            return getattr(self, name)(skipna=skipna, **kwargs)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py:288]: You don't need to pass kwargs here, as there are no possible additional ones for min/max as far as I am aware

**Comment by jorisvandenbossche** [comment]
> What do you think of implementing these reductions at the PandasArray or BaseMaskedArray level (I figure they should probably be used at least for IntegerArray anyways, but if so then might as well make them inheritable by other masked arrays, e.g., BooleanArray)?

Yeah, we should indeed decide on this more generally (for all our internal EAs, so string/boolean/int, I think the older ones already have those methods). And also more broadly for all/most reductions (so also mean, sum, etc). But that can be discussed / done separately from this PR though. Eg one question is if we actually add this to the EA interface (so base class EA), or only to our own EAs. 
Do you want to open an issue for this?

File: pandas/core/arrays/string_.py (hunk 2 above)
Comment (jreback) [reply pandas/core/arrays/string_.py:293]: we don’t do this anywhere else so -1 on skipping the check here (this overhead is tiny anyhow); pls don’t prematurely optimize things 

it makes the code much harder to follow and read

File: pandas/tests/extension/base/reduce.py [hunk 4]
```diff
@@ -25,6 +25,13 @@ class BaseNoReduceTests(BaseReduceTests):
 
     @pytest.mark.parametrize("skipna", [True, False])
     def test_reduce_series_numeric(self, data, all_numeric_reductions, skipna):
+        if isinstance(data, pd.arrays.StringArray) and all_numeric_reductions in [
```
Comment (jreback) [inline pandas/tests/extension/base/reduce.py:28]: hmm, @TomAugspurger, @jbrockmendel  is this the pattern we are using here for skips like this?

File: pandas/core/arrays/string_.py [hunk 5]
```diff
@@ -282,8 +284,25 @@ def astype(self, dtype, copy=True):
         return super().astype(dtype, copy)
 
     def _reduce(self, name, skipna=True, **kwargs):
+        if name in ["min", "max"]:
+            return getattr(self, name)(skipna=skipna)
+
         raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
 
+    def min(self, skipna=True, **kwargs):
```
Comment (jreback) [inline pandas/core/arrays/string_.py:292]: can you type the args and return values

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good.

File: pandas/core/arrays/string_.py [hunk 6]
```diff
@@ -282,8 +284,25 @@ def astype(self, dtype, copy=True):
...
+        nv.validate_min((), kwargs)
+        result = masked_reductions.min(
+            values=self.to_numpy(), mask=self.isna(), skipna=skipna
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_.py:295]: isnt the point of masked_reductions that we _dont_ have to do to_numpy?

File: pandas/tests/extension/base/reduce.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/extension/base/reduce.py:28]: in indexes tests when we have a "this is tested in this other place" comment we usually return/pass instead of pytest.skip.  

It's not a perfect system, but I think of it as a way of distinguishing between "this test is skipped but would be nice to enable" vs "nothing to see here"

File: pandas/core/arrays/string_.py (hunk 6 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py:295]: I was following the annotations there, but taking another look I don't see why values has to be an ndarray. Maybe those types could be updated?

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py:296]: `masked_reductions.py` is purely numpy. You pass it the numpy array parts of the EA. 
This could also be `self._ndarray` instead of `self.to_numpy()`, but both are exactly the same

File: pandas/tests/extension/base/reduce.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/tests/extension/base/reduce.py:28]: Returning None rather than skipping also keeps the test output cleaner, since we print skipped tests.

**Comment by dsaxton** [comment]
> Yeah, we should indeed decide on this more generally (for all our internal EAs, so string/boolean/int, I think the older ones already have those methods). And also more broadly for all/most reductions (so also mean, sum, etc). But that can be discussed / done separately from this PR though.

After annotating the StringArray methods mypy complains about a mismatch between the signatures with PandasArray. I think we could just replace the parent class implementations in this PR and that should still work / be a bit cleaner?

File: pandas/tests/extension/base/reduce.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/reduce.py:28]: We typically refine this for a specific type by overriding this test to have custom behaviour in `extension/test_strings.py`

File: pandas/core/arrays/numpy_.py [hunk 7]
```diff
@@ -349,13 +351,19 @@ def all(self, axis=None, out=None, keepdims=False, skipna=True):
         nv.validate_all((), dict(out=out, keepdims=keepdims))
         return nanops.nanall(self._ndarray, axis=axis, skipna=skipna)
 
-    def min(self, axis=None, out=None, keepdims=False, skipna=True):
```
Comment (TomAugspurger) [inline pandas/core/arrays/numpy_.py:352]: What's the motivation for these changes? Pleasing mypy?

To the extent possible we'd like to keep the signature compatible with ndarray's methods, so that we work well with things like NumPy's dispatch APIs.

Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py:352]: Those keywords are still accepted as kwargs, and now actually are validated (`nv.validate_min`), which was before not the case. 
Those keywords don't do anything for this array class, so IMO it is good to put them in kwargs (we do the same pattern in eg Series.min).

Comment (jreback) [reply pandas/core/arrays/numpy_.py:352]: seems ok to me to allow kwargs; agree we do this elsewhere

File: pandas/tests/arrays/string_/test_string.py [hunk 8]
```diff
@@ -230,6 +230,26 @@ def test_reduce(skipna):
...
+        assert result == expected
+    else:
+        assert result is pd.NA
+
+
+@pytest.mark.parametrize("method", ["min", "max"])
+def test_min_max_numpy(method):
+    arr = pd.Series(["a", "b", "c", None], dtype="string")
+    result = getattr(np, method)(arr)
+    expected = "a" if method == "min" else "c"
```
Comment (jreback) [inline pandas/tests/arrays/string_/test_string.py:249]: why isn't this return the same, IIRC numpy just dispatches back to the array, so this should yield the same as above?

File: pandas/tests/extension/base/reduce.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/extension/base/reduce.py:28]: ok on the test names? @jorisvandenbossche

File: pandas/tests/arrays/string_/test_string.py (hunk 8 above)
Comment (dsaxton) [reply pandas/tests/arrays/string_/test_string.py:249]: It's pretty much the same test, I only separated it out so we're only testing numpy min / max once (no need to parametrize over skipna). I could add a few lines to the test above and remove this if that's less clutter

Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py:249]: I think it is fine (cleaner) to keep the tests separate like this

File: pandas/tests/extension/base/reduce.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/reduce.py:28]: @dsaxton can you move this special case for string to `test_strings.py` ? (and there you an override this test method, to do the correct thing for string dtype)

**Comment by simonjayhawkins** [comment]
i'm seeing a regression here...

```
>>> import numpy as np
>>> import pandas as pd
>>>
>>> pd.__version__
'1.1.0.dev0+1273.gfd6296309'
>>>
>>> arr = pd.array([np.inf, 100])
>>>
>>> np.max(arr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in amax
  File "C:\Users\simon\Anaconda3\envs\pandas-dev\lib\site-packages\numpy\core\fromnumeric.py", line 2667, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "C:\Users\simon\Anaconda3\envs\pandas-dev\lib\site-packages\numpy\core\fromnumeric.py", line 88, in _wrapreduction
    return reduction(axis=axis, out=out, **passkwargs)
  File "C:\Users\simon\pandas\pandas\core\arrays\numpy_.py", line 362, in max
    nv.validate_max((), kwargs)
  File "C:\Users\simon\pandas\pandas\compat\numpy\function.py", line 70, in __call__
    validate_args_and_kwargs(
  File "C:\Users\simon\pandas\pandas\util\_validators.py", line 205, in validate_args_and_kwargs
    validate_kwargs(fname, kwargs, compat_args)
  File "C:\Users\simon\pandas\pandas\util\_validators.py", line 148, in validate_kwargs
    _check_for_invalid_keys(fname, kwargs, compat_args)
  File "C:\Users\simon\pandas\pandas\util\_validators.py", line 122, in _check_for_invalid_keys
    raise TypeError(f"{fname}() got an unexpected keyword argument '{bad_arg}'")
TypeError: max() got an unexpected keyword argument 'axis'
>>>
```

on master
```
>>> import numpy as np
>>> import pandas as pd
>>>
>>> pd.__version__
'1.1.0.dev0+1347.g428791c5e'
>>>
>>> arr = pd.array([np.inf, 100])
>>>
>>> np.max(arr)
inf
>>>
```

**Comment by dsaxton** [comment]
> i'm seeing a regression here...

I think this is caused by validating all the kwargs, any idea what the best approach is here @jorisvandenbossche ? Looks like before only out and keepdims were validated

**Comment by jorisvandenbossche** [comment]
Yeah, that's because the validation functions is not handling `axis` properly

File: pandas/tests/arrays/string_/test_string.py [hunk 9]
```diff
@@ -230,6 +230,26 @@ def test_reduce(skipna):
...
+    if skipna:
+        expected = "a" if method == "min" else "c"
+        assert result == expected
+    else:
+        assert result is pd.NA
+
+
+@pytest.mark.parametrize("method", ["min", "max"])
+def test_min_max_numpy(method):
+    arr = pd.Series(["a", "b", "c", None], dtype="string")
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/string_/test_string.py:247]: so we should also test here for the actual array (and not put in a Series), maybe parametrize over pd.array and pd.Series as constructor

**Review by jreback — APPROVED** [review]
lgtm. @jorisvandenbossche

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @dsaxton !
